### PR TITLE
Support emitting Public API information for records

### PR DESCRIPTION
**Pull request by CyrusNajmabadi**

### CODE REVIEW

File: src/PublicApiAnalyzers/Core/Analyzers/DeclarePublicApiAnalyzer.Impl.cs [hunk 1]
```diff
@@ -174,18 +174,29 @@ private void OnSymbolActionCore(ISymbol symbol, Action<Diagnostic> reportDiagnos
...
                             OnSymbolActionCore(implicitConstructor, reportDiagnostic, isImplicitlyDeclaredConstructor: true, obsoleteAttribute, cancellationToken, explicitLocation: explicitLocation);
                         }
                     }
+
+                    // Ensure that any implicitly declared members of a record are emitted as well.
+                    foreach (var member in namedType.GetMembers())
+                    {
+                        if (IsTrackedAPI(member, cancellationToken) && member is IMethodSymbol { IsImplicitlyDeclared: true } method)
+                        {
+                            if (method.MethodKind is not (MethodKind.Constructor or MethodKind.PropertyGet or MethodKind.PropertySet) ||
```
Comment (sharwell) [inline src/PublicApiAnalyzers/Core/Analyzers/DeclarePublicApiAnalyzer.Impl.cs:193]: Why skip ctor?

File: src/PublicApiAnalyzers/Core/Analyzers/DeclarePublicApiAnalyzer.Impl.cs [hunk 2]
```diff
@@ -732,9 +743,13 @@ internal void ReportMarkedAsRemovedButNotActuallyRemovedApiList(Action<Diagnosti
...
             private bool IsTrackedAPI(ISymbol symbol, CancellationToken cancellationToken)
             {
-                if (symbol is IMethodSymbol methodSymbol && s_ignorableMethodKinds.Contains(methodSymbol.MethodKind))
+                if (symbol is IMethodSymbol methodSymbol)
                 {
-                    return false;
+                    if (s_ignorableMethodKinds.Contains(methodSymbol.MethodKind))
+                        return false;
+
+                    if (methodSymbol is { MethodKind: MethodKind.Ordinary, CanBeReferencedByName: false })
```
Comment (sharwell) [inline src/PublicApiAnalyzers/Core/Analyzers/DeclarePublicApiAnalyzer.Impl.cs:757]: ❓ Shouldn't we include the `<Clone>` method? Users might not realize that converting a record back to a non-record will always be a breaking change.

File: src/PublicApiAnalyzers/Core/Analyzers/DeclarePublicApiAnalyzer.Impl.cs (hunk 1 above)
Comment (CyrusNajmabadi) [reply src/PublicApiAnalyzers/Core/Analyzers/DeclarePublicApiAnalyzer.Impl.cs]: oopsy :).  fixed.

File: src/PublicApiAnalyzers/Core/Analyzers/DeclarePublicApiAnalyzer.Impl.cs (hunk 2 above)
Comment (CyrusNajmabadi) [reply src/PublicApiAnalyzers/Core/Analyzers/DeclarePublicApiAnalyzer.Impl.cs:760]: hrmmmmmmmmmmmmmmmmmmmmmmmmmmmmmmmmmmmmmmm.  ok.  that is a compellign point.

**Review by sharwell — APPROVED** [review]

**Comment by CyrusNajmabadi** [comment]
Thanks @sharwell   can you update Roslyn to use this when you have moment?

**Pull request merged**